src/Grpc/JsonTranscoding/test/Microsoft.AspNetCore.Grpc.JsonTranscoding.Tests/JsonTranscodingRouteAdapterTests.cs

```diff
@@ -0,0 +1,223 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Grpc.Shared;
+using Microsoft.AspNetCore.Grpc.JsonTranscoding.Internal;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.FileSystemGlobbing.Internal;
+
+namespace Microsoft.AspNetCore.Grpc.JsonTranscoding.Tests;
+
+public class JsonTranscodingRouteAdapterTests
+{
+    [Fact]
+    public void ParseMultipleVariables()
+    {
+        var pattern = HttpRoutePattern.Parse("/shelves/{shelf}/books/{book}")!;
+        var adapter = JsonTranscodingRouteAdapter.Parse(pattern);
+
+        Assert.Equal("/shelves/{shelf}/books/{book}", adapter.ResolvedRouteTemplate);
+        Assert.Empty(adapter.RewriteVariableActions);
+    }
+
+    [Fact]
+    public void ParseComplexVariable()
+    {
+        var route = HttpRoutePattern.Parse("/v1/{book.name=shelves/*/books/*}")!;
+        var adapter = JsonTranscodingRouteAdapter.Parse(route);
+
+        Assert.Equal("/v1/shelves/{__Complex_book.name_2}/books/{__Complex_book.name_4}", adapter.ResolvedRouteTemplate);
+        Assert.Single(adapter.RewriteVariableActions);
+
+        var httpContext = new DefaultHttpContext();
+        httpContext.Request.RouteValues = new RouteValueDictionary
+        {
+            { "__Complex_book.name_2", "first" },
```

Comment:
I'm a bit confused by this, aren't these route values provided by the user? The user is going to use "book.name" not some internal encoded string like this.

Reply:
Look at the code immediately below your comment. The RewriteVariableAction is called and reconstructs these internal values into a value set to `book.name`.

I've also made the comment at the top of JsonTranscodingRouteAdapter more detailed.

Reply:
Figured it out by looking at the test logs, we are constructing the routing template for the routing middleware which is what looks at the route provided by the user and matches + constructs the `RouteValueDictionary`.
e.g. we generate `"/v1/{name.last_name}/complex_greeter/{**__Complex_name.first_name_3:regex(b/c/d/two$)}"` and pass that to routing for matching.